Source/wayland/WaylandServer+Cursor.m

```diff
@@ -0,0 +1,564 @@
+                    edges = XDG_TOPLEVEL_RESIZE_EDGE_BOTTOM_LEFT;
+                } else if (eventLocation.x > resizeBarRect.size.width - 30) {
+                    edges = XDG_TOPLEVEL_RESIZE_EDGE_BOTTOM_RIGHT;
+                } else if (eventLocation.x < 29) {
+                    edges = XDG_TOPLEVEL_RESIZE_EDGE_BOTTOM_LEFT;
+                } else {
+                    edges = XDG_TOPLEVEL_RESIZE_EDGE_BOTTOM;
+                }
+
+                if(state == WL_POINTER_BUTTON_STATE_PRESSED) {
+                    xdg_toplevel_resize(window->toplevel, wlconfig->seat, serial, edges);
+                    window->resizing = YES;
+                    return;
+                } else {
+                    window->resizing = NO;
+                }
+            }
+        }
+    } // endif window->toplevel
+
+
+    if (state == WL_POINTER_BUTTON_STATE_PRESSED) {
+	if (button == wlconfig->pointer.last_click_button &&
+	    time - wlconfig->pointer.last_click_time < 300 &&
+	    abs(wlconfig->pointer.x - wlconfig->pointer.last_click_x) < 3 &&
+	    abs(wlconfig->pointer.y - wlconfig->pointer.last_click_y) < 3) {
+	    wlconfig->pointer.last_click_time = 0;
+	    clickCount++;
+	} else {
+	    NSDebugLog(@"handle_button MISS: b=%d t=%d x=%f y=%f", button, time, wlconfig->pointer.x, wlconfig->pointer.y);
+	    wlconfig->pointer.last_click_button = button;
+	    wlconfig->pointer.last_click_time = time;
+	    wlconfig->pointer.last_click_x = wlconfig->pointer.x;
+	    wlconfig->pointer.last_click_y = wlconfig->pointer.y;
+	}
+
+	switch (button) {
+	case BTN_LEFT:
+	    eventType = NSLeftMouseDown;
+	    break;
+	case BTN_RIGHT:
+	    eventType = NSRightMouseDown;
+	    break;
+	case BTN_MIDDLE:
+	    eventType = NSOtherMouseDown;
+	    break;
+        // TODO: handle BTN_SIDE, BTN_EXTRA, BTN_FORWARD, BTN_BACK and other
+        // constants in libinput.
+        // We may just want to send NSOtherMouseDown and populate buttonNumber
+        // with the libinput constant?
+	}
+	wlconfig->pointer.serial = serial;
+    } else if (state == WL_POINTER_BUTTON_STATE_RELEASED) {
+	switch (button) {
+	case BTN_LEFT:
+	    eventType = NSLeftMouseUp;
+	    break;
+	case BTN_RIGHT:
+	    eventType = NSRightMouseUp;
+	    break;
+	case BTN_MIDDLE:
+	    eventType = NSOtherMouseUp;
+	    break;
+	}
+	wlconfig->pointer.serial = 0;
+    } else {
+      return;
+    }
+
+    /* FIXME: unlike in _motion and _axis handlers, the argument used in _button
+       is the "serial" of the event, not passed and unavailable in _motion and
+       _axis handlers. Is it allowed to pass "serial" as the eventNumber: in
+       _button handler, but "time" as the eventNumber: in the _motion and _axis
+       handlers? */
+    tick = serial;
+
+    NSDebugLog(@"sending pointer event at: %fx%f, window=%d", wlconfig->pointer.x, wlconfig->pointer.y, window->window_id);
+
+    /* FIXME: X11 backend uses the XGetPointerMapping()-returned values from
```

Comment:
We also support the version of the event without the button number. Just use that if you prefer. The more important bit is to set the event type correctly as for the first three buttons we already have the information there.

Reply:
I think that the current code structure provides a good separation of concerns: Cursors, Keyboard, Layer/Xdg protocols, Seat and Output. Starting from this we could start collecting feedback on the separate areas and tackle them 1 by 1

Reply:
I don't see how your reply is in any way related to my comment. You probably wanted to explain that this code is not yours, you just moved it into a separate file.

Reply:
yes, the comment is more related to the general feedback that the PR is very large to be reviewed. Anyway yes, the Cursor code has only been moved in a new file